runtime/registry.go

```diff
@@ -502,3 +503,54 @@ func (r *registryCache) ensureProjectParser(ctx context.Context, instanceID stri
 		r.logger.Error("could not create project parser", zap.Error(err), zap.String("instance_id", instanceID), observability.ZapCtx(ctx))
 	}
 }
+
+// instanceHeartbeatEmitter emits instance heartbeat events periodically after specified duration until closed.
+type instanceHeartbeatEmitter struct {
+	rt       *Runtime
+	duration time.Duration
+
+	closeChan chan struct{}
+}
+
+func newInstanceHeartbeatEmitter(rt *Runtime, duration time.Duration) *instanceHeartbeatEmitter {
+	return &instanceHeartbeatEmitter{
+		rt:       rt,
+		duration: duration,
+	}
+}
+
+func (i *instanceHeartbeatEmitter) emit() {
```

Comment:
This doesn't really require a separate struct – it can just be a function on `registryCache`, like `go r.emitHeartbeats()` in `newRegistryCache`. 

The registry cache already has a `baseCtx` which you can use instead of `closeChan` for cancellation.

Reply:
Cool. Added in `init` since that's where cache is populated.